### PR TITLE
Allow GeneratorParam<std::string>

### PR DESCRIPTION
**Pull request by asaadaldien**
This will allow us to pass and AOT arbitrary serialized objects (e.g onnx protobuffs).

### CODE REVIEW

**Comment by abadams** [comment]
Deferring to @steven-johnson to check if this is the right way to add std::string

File: test/generator/string_param_generator.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#include "Halide.h"
```
Comment (steven-johnson) [inline test/generator/string_param_generator.cpp]: The test needs to be augmented to ensure that setting the generatorparam in the Makefile (and also the CMake file) works correctly.

File: src/Generator.h [hunk 2]
```diff
@@ -939,6 +945,30 @@ class GeneratorParam : public Internal::GeneratorParamImplBase<T> {
         : Internal::GeneratorParamImplBase<T>(name, value) {}
 };
 
+// Explicitly initiate std::string type outside to avoid constructor duplicate signature.
```
Comment (steven-johnson) [inline src/Generator.h]: This works but is dramatically different in style from how all of the other GeneratorParam subtypes are implemented; mimicking the existing style would arguably be better, though probably more verbose.

Comment (steven-johnson) [reply src/Generator.h]: The comment block above here (line 898 or so) is now out of date and needs updating.

File: test/generator/string_param_generator.cpp (hunk 1 above)
Comment (steven-johnson) [reply test/generator/string_param_generator.cpp]: We also need a test to verify that GeneratorParam<string> works correctly with Stubs; the easiest way to do that is add one to stubtest_generator.cpp and then to reference it in stubtest_aottest.cpp and also in stubuser_generator.cpp

File: test/generator/string_param_generator.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+class StringParam : public Halide::Generator<StringParam> {
+public:
+    GeneratorParam<std::string> param{"param", "add_one"};
+
+    Input<Buffer<float>> input{ "input", 1 };
+    Output<Buffer<float>> output{ "output", 1 };
+
+    void generate() {
+        Var x;
+        if (param.value() == "add_one") {
```
Comment (steven-johnson) [inline test/generator/string_param_generator.cpp]: I realize you want a simple test, but this is a very bad example that I fear people will mimic in new code; in this case, using a GeneratorParam<bool> or GeneratorParam<EnumType>, etc would be vastly more appropriate. Can we come up with a test example that actually *requires* the string type, and none of the other existing types will reasonably substitute?

Comment (abadams) [reply test/generator/string_param_generator.cpp]: How about an RPN description of the computation to perform e.g: "3 x + y +". That just requires Halide::Internal::split_string to break it into tokens and a stack (`vector<Expr>`) to parse into some Expr to use as the RHS of a Func.

Comment (steven-johnson) [reply test/generator/string_param_generator.cpp]: SGTM

Comment (asaadaldien) [reply test/generator/string_param_generator.cpp]: I agree with @steven-johnson its a bad example, I added it just want to make sure buildbots are happy with this change. @abadams example is perfect will update the test accordingly.

**Comment by slomp** [comment]
Thaks for considering and implementing this feature 👍

**Comment by steven-johnson** [comment]
LGTM; only piece missing is that we should verify that setting a string value in Make / CMake files works correctly.

File: test/generator/stubuser_generator.cpp [hunk 4]
```diff
@@ -54,6 +54,7 @@ class StubUser : public Halide::Generator<StubUser> {
         gp.untyped_buffer_output_type = int32_buffer_output.type();
         gp.intermediate_level.set(LoopLevel(calculated_output, Var("y")));
         gp.vectorize = true;
+        gp.str_param = "2 * x y +";
```
Comment (abadams) [inline test/generator/stubuser_generator.cpp]: Nit: this particular RPN expression would underflow the stack.

Comment (asaadaldien) [reply test/generator/stubuser_generator.cpp]: Ops!

**Comment by steven-johnson** [comment]
LGTM when everything is green. Thanks for the work!

**Comment by asaadaldien** [comment]
Thanks for the review!

**Comment by asaadaldien** [comment]
CMake removes double quote"" from argument value! e.g: 
```
  halide_define_aot_test(string_param
                         GENERATOR_ARGS rpn_expr="5 x * y +")
```

Generator is called with `rpn_expr=5\ x\ *\ y\ +`.

**Comment by steven-johnson** [comment]
Ah, CMake. The gift that keeps on giving.

**Pull request merged**